sdk/eventhubs/azure-messaging-eventhubs/inc/azure/messaging/eventhubs/processor.hpp

```diff
@@ -89,127 +94,206 @@ namespace Azure { namespace Messaging { namespace EventHubs {
-     * @param consumers The map of partition id to partition client.
-     * @param context The context to control the request lifetime.
+     * @param context The context to control the request lifetime of the processor. Cancelling this
+     * context will stop the processor from running.
+     *
+     * @remark This function starts the processor running in a new thread.
      */
-    void Dispatch(
-        Models::EventHubProperties const& eventHubProperties,
-        ConsumersType& consumers,
-        Core::Context const& context)
-    {
-      std::vector<Models::Ownership> ownerships
-          = m_loadBalancer->LoadBalance(eventHubProperties.PartitionIds, context);
+    void Start(Azure::Core::Context const& context = {});
 
-      std::map<std::string, Models::Checkpoint> checkpoints = GetCheckpointsMap(context);
+    /** @brief Stops a running processor.
+     *
+     * @remark This function stops the processor. If the Start method has been called, it will wait
+     * for the thread to complete.
+     */
+    void Stop();
 
-      for (auto const& ownership : ownerships)
+    /** @brief Closes the processor and cancels any current operations.
+     *
+     */
+    void Close(Core::Context const& context = {})
+    {
+      if (m_isRunning)
       {
-        AddPartitionClient(ownership, checkpoints, consumers);
+        throw std::runtime_error("cannot close a processor that is running");
       }
+
+      // Drain the partition clients queue.
+      for (;;)
+      {
+        auto client = m_nextPartitionClients.TryRemove();
+        if (client)
+        {
+          client->Close();
+        }
+        else
+        {
+          break;
+        }
+      }
+      (void)context;
     }
 
-    /** @brief Closes the processor and cancels any current operations.
+  private:
+    /** Representation of Go channel construct.
+     *
+     * A channel represents a size limited queue, where items can be inserted and removed. If there
+     * are no items in the queue, the caller will block until an item is inserted into the queue.
      *
+     * @tparam T The type of the items in the queue.
      *
      */
-    void Close()
-    {
-      for (auto& consumer : m_nextPartitionClients)
+    template <class T> class Channel {
+    public:
+      Channel() : m_maximumDepth{0} {}
+
+      ~Channel()
       {
-        consumer->Close();
+        Azure::Core::Diagnostics::_internal::Log::Stream(
+            Azure::Core::Diagnostics::Logger::Level::Verbose)
+            << "~Channel. Currently depth is " << m_channelDepth << " and maximum depth is "
+            << m_maximumDepth;
+        Azure::Core::Diagnostics::_internal::Log::Stream(
+            Azure::Core::Diagnostics::Logger::Level::Verbose)
+            << "Clear channel queue.";
+        while (m_channelDepth > 0)
+        {
+          auto value = m_channelQueue.TryWaitForResult();
+          if (value)
+          {
+            m_channelDepth -= 1;
```

Comment:
This is the destructor for the Channel. By definition there cannot be any other code referencing the Channel object, so there is no need to acquire a lock.

Reply:
Note that there IS a potential issue in the Remove method because we cannot take the channel lock across the call to WaitForResult. The window is small but real.